content/4_1_couchbase.tex

```diff
@@ -0,0 +1,165 @@
+
+\begin{figure}
+    \centering
+    \includegraphics[width=\textwidth]{img/couchbaseMonitorOps.jpg}
+    \caption{Monitoring the operations per second}
+    \label{figure:ops}
+\end{figure}
+
+\autoref{figure:ops} shows two spikes in operations per second in the last hour, which timestamps correspond to the queries executed before.
+
+\begin{figure}[H]
+    \centering
+    \includegraphics[width=\textwidth]{img/couchbaseMonitor.jpg}
+    \caption{Monitoring metrics of an individual bucket in Couchbase}
+    \label{figure:Monitoring}
+\end{figure}
+
+\autoref{figure:Monitoring} shows all the graphs available for individual buckets. All in all, over forty metrics can be viewed and used to assess the health and performance of the system.
+
+\subsection{CAP Theorem}
+From a CAP-Theorem perspective Couchbase can be operated as a CP or CA system. By default it will act as CP system. When a Node fails some data will be temporarily unavailable for writes until the replica is set to active. However, reads will be processed regularly by the replicas available in the cluster. If needed, it can be tuned to act as a CA system for example through auto-failover. But this is only half the story and only applies if Couchbase is run in a single cluster setting. With a multiple cluster setting, also known as Cross Data Center Replication (XDCR), it will act as AP system. Since in this setting any Cluster can be written to, problems arise when the same data is modified from different locations. In this case Couchbase provides different configurations for conflict resolution (for example through timestamp or revision ID), eventually leading to consistency between multiple clusters. \parencite{couchbaseCAP}
+
+\subsection{SDK}
+\label{section:SDK}
+To use Couchbase productively in a development environment this last section shall discuss the available software development kit and give some examples on how to use it to perform basic \gls{crud} operations on the cluster's data.\\
+First of all, Couchbase offers SDKs for several popular programming languages and runtime environments inlcuding C, Go, Java, .NET, Node.js, PHP and Python. \parencite{couchbaseSDK} Since it would be out of scope to give examples for all of them, this section will limit the examples to the Node.js SDK, but the information given will be easily transferable to other languages as well.
+
+First, a connection to the Couchbase Cluster needs to be established. Afterwards, authentication is necessary to select the desired bucket. Once there is a bucket object it can be used to send queries. The described steps can be seen in Listing \ref{couchbaseAuthAndBucket}.\\
+\begin{listing}[ht]
+\begin{minted}{js}
+const cluster = new couchbase.Cluster('localhost:8091');
+cluster.authenticate('user', 'password');
+const bucket = cluster.openBucket('demo-bucket');
+\end{minted}
+\caption{Connect to Couchbase}
+\label{couchbaseAuthAndBucket}
+\end{listing}
+
+\newpage
```

Comment:
No offense, but since it's our chapter and doesn't affect the layout of others I won't change this.

Reply:
I disagree. We want to have a book that looks nice in the end. That's why you use LaTeX, to make it look professional, and that should also apply to the code. LaTeX just knows what is right in most cases. For example, I tested the code without this newpage here, and it still made a new page because there has not been enough place anymore. But if I changed the layout (e.g. other paper size, other page boundaries, ...), this would break the layout. So, please try to produce good LaTeX code.

Reply:
If you want to fullfill your own standards that's fine. But I won't spent more time on fulfilling your requirements for a book that is not graded and that most of us will probably never look at again. Neither I nor our group will spent much more time on this. We just have better things to do. Pull or don't pull.

Kind Regards,
Group Couchbase